src/functional/ffront/type_info.py

```diff
@@ -52,17 +52,32 @@ def _is_zero_dim_field(field: ts.TypeSpec) -> bool:
 
 
 def promote_zero_dims(
-    args: list[ts.TypeSpec], function_type: ts.FieldOperatorType | ts.ProgramType
+    args: list[ts.TypeSpec], function_type: ts.FieldOperatorType | ts.ProgramType | ts.FunctionType
 ) -> list:
     """Promote arg types to zero dimensional fields if compatible and required by function signature."""
-    new_args = args.copy()
+    new_args = args
     for arg_i, arg in enumerate(args):
-        def_type = function_type.definition.args[arg_i]
-        if _is_zero_dim_field(def_type) and is_number(arg):
-            if extract_dtype(def_type) == extract_dtype(arg):
-                new_args[arg_i] = def_type
-            else:
-                raise GTTypeError(f"{arg} is not compatible with {def_type}.")
+        def_type = (
+            function_type.args[arg_i]
+            if isinstance(function_type, ts.FunctionType)
+            else function_type.definition.args[arg_i]
+        )
+
+        def _as_field(arg: ts.TypeSpec, path: tuple):
+            try:
+                def_t = reduce(lambda type_, idx: type_.types[idx], path, def_type)  # noqa: B023
```

Comment:
```suggestion
                el_def_type = reduce(lambda type_, idx: type_.types[idx], path, def_type)  # noqa: B023
```
How about `el_def_type`? Or alternatively rename the outer `def_type` to `outer_def_type`. Just `def_t` however is somewhat against our variable naming conventions.